rosetta/docs/GPU_performance.md

```diff
@@ -0,0 +1,131 @@
+
+
+## General CUDA/NCCL flags 
+
+### CUDA Flags
+
+The following flag restricts cuda queues to 1 and is useful when a strict ordering of operations is required to achieve best performance. This is recommended to achieve good performance with latency hiding optimizations with asynchronous collectives.
+- CUDA_DEVICE_MAX_CONNECTIONS=1
+  
+### NCCL flags 
+
+See [Environment Variables] (https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html) for more details.
+- NCCL_LL128_BUFFSIZE: -2
+- NCCL_LL_BUFFSIZE: -2
+- NCCL_PROTO: SIMPLE,LL,LL128
+
+
+## XLA flags to enable Latency Hiding Scheduler, and asynchronous collective communication
+
+To achieve communication computation overlap for models in JAX/XLA, we must enable Latency Hiding Scheduler and enable asynchronous communications. 
+
+To enable latency hiding optimizations with XLA, turn on the following flag: 
+
+- --xla_gpu_enable_latency_hiding_scheduler=true 
+
+To enable asynchronous communication for all collectives, the following is recommended, and is set by default in XLA :
+
+- --xla_gpu_enable_async_collectives=true
+- --xla_gpu_enable_highest_priority_async_stream=true
+
+For more fine-grained control over which collectives should be asynchronous or not, please use: 
+
+- --xla_gpu_enable_async_all_reduce=<>
+- --xla_gpu_enable_async_all_gather=<>
+- --xla_gpu_enable_async_reduce_scatter=<> 
+- --xla_gpu_enable_async_collective_permute=<>
+
+
+### Enable Optimizations for FSDP communication 
+
+With FSDP in JAX/XLA, there are additional optimizations of 
+
+- scan loop unrolling and loop double buffering 
+    - --xla_gpu_enable_while_loop_double_buffering=true
+      
+- optimized pipelining of all-gather and reduce-scatter for latency hiding in FSDP
+    - --xla_gpu_enable_pipelined_all_gather=true
+    - --xla_gpu_enable_pipelined_reduce_scatter=true
+    - --xla_gpu_enable_pipelined_all_reduce=true 
+    - --xla_gpu_enable_pipelined_collectives=false // overrides the above
+      
+- combining tensors that are sharded along different dimensions. Within a transformer layer, tensors can be sharded row-wise or column-wise and by default XLA will generate multiple collective calls for tensors sharded along different dimensions. The following optimization flags combine all tensors shardings, and map them to a group NCCL call that has a large commulative size and achieves high communication efficiency. 
+    - --xla_gpu_enable_all_gather_combine_by_dim=false
+    - --xla_gpu_enable_reduce_scatter_combine_by_dim=false
+      
+- Combine threshold values in XLA that determine when an all-gather (AG) or reduce-scatter (RS) is triggered. We want to set these values to be atleast as large as the size of weights (AG) or gradients (RS) in a single transformer layer. For example, LLAMA2-7B with BF16 weights and gradients, we have 32 transformer layers => each layer has ~218M weights => one would want to set these thresholds to atleast 436MB. 
```

Comment:
```suggestion
- Combine threshold values in XLA that determine when an all-gather (AG) or reduce-scatter (RS) is triggered. We want to set these values to be at least as large as the size of weights (AG) or gradients (RS) in a single transformer layer. For example, LLAMA2-7B with BF16 weights and gradients, we have 32 transformer layers => each layer has ~218M weights => one would want to set these thresholds to at least 436MB. This allow to control the communication size, bigger communication are frequently more efficient, but this can limit the communication/computation overlap.
```